master/internal/user/service.go

```diff
@@ -291,6 +291,76 @@ func (s *Service) patchUser(c echo.Context) (interface{}, error) {
 	}, nil
 }
 
+func (s *Service) patchUsername(c echo.Context) (interface{}, error) {
+	type (
+		request struct {
+			NewUsername *string `json:"username,omitempty"`
+		}
+		response struct {
+			message string
+		}
+	)
+
+	body, err := ioutil.ReadAll(c.Request().Body)
+	if err != nil {
+		return nil, err
+	}
+
+	args := struct {
+		Username string `path:"username"`
+	}{}
+	if err = api.BindArgs(&args, c); err != nil {
+		return nil, err
+	}
+
+	var params request
+	if err = json.Unmarshal(body, &params); err != nil {
+		malformedRequestError := echo.NewHTTPError(http.StatusBadRequest, "bad request")
+		return nil, malformedRequestError
+	}
+
+	forbiddenError := echo.NewHTTPError(http.StatusForbidden)
+	authenticatedUser := c.(*context.DetContext).MustGetUser()
+	user, err := s.db.UserByUsername(args.Username)
+	switch err {
+	case nil:
+	case db.ErrNotFound:
+		if authenticatedUser.Admin {
+			return nil, echo.NewHTTPError(
+				http.StatusBadRequest,
+				fmt.Sprintf("failed to get user '%s'", args.Username))
+		}
+		return nil, forbiddenError
+	default:
+		return nil, err
+	}
+
+	if params.NewUsername == nil {
+		malformedRequestError := echo.NewHTTPError(http.StatusBadRequest, "username is required")
+		return nil, malformedRequestError
+	}
+
+	switch u, uErr := s.db.UserByUsername(*params.NewUsername); {
+	case uErr == db.ErrNotFound:
+	case uErr != nil:
+		return nil, uErr
+	case u != nil:
+		return nil, echo.NewHTTPError(http.StatusBadRequest, "username is taken")
+	}
+
+	if !authenticatedUser.Admin {
```

Comment:
non-blocking: We could move this check up above? If it fails no need to hit the DB. Also it would slightly simplify some logic above, perhaps?